transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,181 @@
+
+    private final ByteBufAllocator alloc;
+    private final ArrayDeque<Object> promisesAndBufs = new ArrayDeque<Object>();
+    private int readableBytes;
+
+    public CoalescingBufferQueue(ByteBufAllocator alloc) {
+        this.alloc = ObjectUtil.checkNotNull(alloc, "alloc");
+    }
+
+    /**
+     * Add a buffer to the end of the queue.
+     */
+    public void add(ByteBuf buf) {
+        promisesAndBufs.add(ObjectUtil.checkNotNull(buf, "buff"));
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Add a buffer to the end of the queue and associate a promise with it that should be completed when
+     * all it's bytes have been consumed from the queue and written.
+     * @param buf to add to the tail of the queue
+     * @param promise to complete when all the bytes have been consumed and written, can be void.
+     */
+    public void add(ByteBuf buf, ChannelPromise promise) {
+        // buffers are added before promises so that we naturally 'consume' the entire buffer during removal
+        // before we complete it's promise.
+        ObjectUtil.checkNotNull(buf, "buf");
+        ObjectUtil.checkNotNull(promise, "promise");
+        promisesAndBufs.add(buf);
+        if (!promise.isVoid()) {
+            promisesAndBufs.add(promise);
+        }
+        readableBytes += buf.readableBytes();
+    }
+
+    /**
+     * Remove a {@link ByteBuf} from the queue with the specified number of bytes. Any added buffer who's bytes are
+     * fully consumed during removal will have it's promise completed when the passed aggregate {@link ChannelPromise}
+     * completes.
+     *
+     * @param bytes the maximum number of readable bytes in the returned {@link ByteBuf}.
+     * @param aggregatePromise used to aggregate the promises for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
+        if (bytes < 0) {
+            throw new IllegalArgumentException("bytes cannot be less than 0 : " + bytes);
+        }
+        ObjectUtil.checkNotNull(aggregatePromise, "aggregatePromise");
+
+        // Use isEmpty rather than readableBytes==0 as we may have a promise associated with an empty buffer.
+        if (promisesAndBufs.isEmpty()) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        ByteBuf toReturn = null;
+        CompositeByteBuf composite = null;
+        int originalBytes = bytes;
+        for (;;) {
+            Object entry = promisesAndBufs.poll();
+            if (entry == null) {
+                break;
+            } else if (entry instanceof ChannelPromise) {
+                aggregatePromise.addListener(new ChannelPromiseNotifier((ChannelPromise) entry));
+            } else if (bytes > 0) {
+                ByteBuf entryBuffer = (ByteBuf) entry;
+                int entryReadableBytes = entryBuffer.readableBytes();
+                if (toReturn == null) {
```

Comment:
This will only ever be true on the first iteration. Consider unrolling the loop a bit and pull this out of the loop. At most I think you need to check the first 2 entries before getting into the proper loop.